### PR TITLE
[ttnn] batch_norm RunningStatistics: drop redundant custom compute_program_hash

### PR DESCRIPTION
**Pull request by dgomezTT**
RunningStatistics::compute_program_hash hashed the whole attributes struct
(momentum/memory_config/compute_kernel_config/input_dtype/dtype) + each tensor's
TensorSpec plus a redundant padded_shape (derived from TensorSpec) and a redundant
optional re-encoding. The framework default hashes the whole attrs struct +
tensor_args, keying on the same (logical-shape) distinctions.

Adds test_batch_norm_program_cache.py (via batch_norm training path) pinning
that reuse adds no entries and shape/momentum re-key, identical before and after;
functional test_batch_norm.py (1539) green.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:dgomez/drop-hash-batch_norm)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:dgomez/drop-hash-batch_norm)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+Program-cache regression tests for the RunningStatistics device op, exercised via
+ttnn.batch_norm(training=True) (which updates running_mean/running_var through it).
+
+Pins the cache-keying granularity so it can be verified BEFORE and AFTER removing
+RunningStatistics::compute_program_hash. The custom hash keyed on the whole attributes struct
+(momentum/memory_config/compute_kernel_config/input_dtype/dtype) + each tensor's TensorSpec plus a
+redundant padded_shape (derived from tensor_spec) and a redundant optional re-encoding. The framework
+default hashes the whole attrs struct + tensor_args, i.e. the same distinctions and the same
+logical-shape keying, so the TOTAL program-cache entry count for a batch_norm(training=True) call is
+unchanged.
```
Comment (fplavecTT) [inline tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: This describes why this test is good to test this PR (AI loves to write comments like that), which actually belongs into PR description. Comments in test source files should explain why we want this test to be part of our test suite, which is usually a different reason (though it can be motivated by a given change). AI needs to be nudged to write comments in this fashion.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    device.enable_program_cache()
+    yield
+    device.disable_and_clear_program_cache()
+
+
+def run_bn_training(device, input_shape, momentum, seed=0):
+    """ttnn.batch_norm(training=True) inside the cache counter (updates running stats via RunningStatistics)."""
+    torch.manual_seed(seed)
+    channels = input_shape[1]
+    inp = ttnn.from_torch(torch.rand(input_shape, dtype=torch.bfloat16) * 5 + 5, device=device, layout=ttnn.TILE_LAYOUT)
```
Comment (fplavecTT) [inline tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Given that these tests aren't doing anything with the values, can we drop the magic constants (5, 5, 6, 4, 16, 4), since they don't seem to serve any purpose?

**Review by fplavecTT — COMMENTED** [review]
Please address or comment on AI review.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        layout=ttnn.TILE_LAYOUT,
+    )
+    var = ttnn.from_torch(
+        torch.rand(channels, dtype=torch.bfloat16).view(1, channels, 1, 1) * 16 + 4,
+        device=device,
+        layout=ttnn.TILE_LAYOUT,
+    )
+    with device.cache_entries_counter.measure():
+        out = ttnn.batch_norm(inp, running_mean=mean, running_var=var, training=True, eps=1e-5, momentum=momentum)
+    return ttnn.to_torch(out)
```
[the remark this answers is not in the document]
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Done. Every run now reads back the in-place-updated running_mean/running_var and validates them against a torch.nn.functional.batch_norm reference (initial stats placed well above the batch statistics so a cache hit that leaves a call's buffers un-updated is detectable). The RunningStatistics prim has no Python binding, so it stays exercised through ttnn.batch_norm(training=True), its only public entry point.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        device=device,
+        layout=ttnn.TILE_LAYOUT,
+    )
+    var = ttnn.from_torch(
+        torch.rand(channels, dtype=torch.bfloat16).view(1, channels, 1, 1) * 16 + 4,
+        device=device,
+        layout=ttnn.TILE_LAYOUT,
+    )
+    with device.cache_entries_counter.measure():
+        out = ttnn.batch_norm(inp, running_mean=mean, running_var=var, training=True, eps=1e-5, momentum=momentum)
```
[the remark this answers is not in the document]
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Added test_bn_cache_optional_stat_presence, which runs (mean+var), (mean only), and (var only) via training-mode batch_norm. It asserts each new presence combination adds a distinct cache entry and a repeated combination reuses it. The upstream reduction/batch-norm programs are identical across combos (same input), so the growth is attributable to RunningStatistics' presence compile-time args. Present combos also validate the resulting update.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    run_bn_training(device, (1, 4, 32, 32), momentum=0.1, seed=1)
+    assert device.cache_entries_counter.total == n_after_first  # reuse: no growth
+
+
+def test_bn_cache_miss_different_shape(device, isolate_program_cache):
+    """Different input shape -> more entries than a pure reuse."""
+    run_bn_training(device, (1, 4, 32, 32), momentum=0.1)
+    n_after_first = device.cache_entries_counter.total
+    run_bn_training(device, (1, 4, 64, 64), momentum=0.1)
+    assert device.cache_entries_counter.total > n_after_first
```
[the remark this answers is not in the document]
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Fixed: the miss test now varies C (4 -> 8) instead of H/W, so the RunningStatistics per-channel tensor specs (1,C,1,1) actually change and re-key the program. Renamed to test_bn_cache_miss_different_channels, with a comment noting H/W alone would not re-key RunningStatistics.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py (hunk 1 above)
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Rewrote the module and per-test comments to state the invariants the suite upholds (RunningStatistics cache-key granularity, and the running-stats side effect landing on each call's own buffers), rather than the PR motivation. The before/after-hash framing is gone.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py (hunk 2 above)
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Dropped the magic constants. The tests now use the values meaningfully: input is torch.rand in [0,1) and the running stats start from named INIT_RUNNING_MEAN/INIT_RUNNING_VAR constants, chosen so the momentum-weighted update is observable and validated against the torch reference.

**Review by edwinleeTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    # running_var absent -> distinct presence bits -> new program.
+    _run_bn_training(device, BASE_SHAPE, momentum=0.1, with_mean=True, with_var=False, seed=2)
+    n_mean_only = device.cache_entries_counter.total
+    assert n_mean_only > n_both
+    _run_bn_training(device, BASE_SHAPE, momentum=0.1, with_mean=True, with_var=False, seed=3)
+    assert device.cache_entries_counter.total == n_mean_only  # same combo -> reuse
+
+    # running_mean absent (var present) -> another distinct combination -> new program.
+    _run_bn_training(device, BASE_SHAPE, momentum=0.1, with_mean=False, with_var=True, seed=4)
+    assert device.cache_entries_counter.total > n_mean_only
```
Comment (fplavecTT) [inline tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Why do we not check reuse for this case here as we do for other cases?

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    n_after_first = device.cache_entries_counter.total
+    _run_bn_training(device, BASE_SHAPE, momentum=0.5)
+    assert device.cache_entries_counter.total > n_after_first
+
+
+def test_bn_cache_optional_stat_presence(device, isolate_program_cache):
+    """running_mean/running_var presence is baked into the RunningStatistics program, so each
+    present/absent combination must key a distinct program while a repeated combination reuses it.
+    Only RunningStatistics depends on this presence, so the entry growth is attributable to it
+    (the upstream reduction/batch-norm programs are identical across combinations)."""
```
Comment (fplavecTT) [inline tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py:170]: `with_mean=False, with_var=False` combination is missing from this test.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py (hunk 6 above)
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py]: Good catch — the var-only combo now has a repeat run asserting reuse (`total == n_var_only`), consistent with the other combinations. (in bb0fa90)

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py (hunk 7 above)
Comment (dgomezTT) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm_program_cache.py:170]: Added. In training mode `running_statistics` always runs (even with both stats absent — batch_norm.cpp:129), so `(with_mean=False, with_var=False)` is a valid fourth presence combo; it now asserts a distinct program keys and a repeat reuses. (in bb0fa90)

**Comment by dgomezTT** [comment]
Pushed bb0fa90 addressing the review:
- Optional-stat presence test now covers all four `running_mean`/`running_var` combinations (incl. `(absent, absent)`), and every present/absent combo — including var-only — asserts both a distinct cache key and reuse on repeat.
- Closed a coverage gap not in the comments: the suite ran only bfloat16 and never exercised the RunningStatistics compile-time `any_float32` branch (which selects a different compute kernel). Added `test_bn_cache_miss_different_dtype` pinning that bf16 vs fp32 key distinct programs and a repeated dtype reuses — the dtype keying the default hash must preserve now that the custom hash is gone.

All 5 tests green locally on BH.

**Review by bbradelTT — APPROVED** [review]

**Review by fplavecTT — APPROVED** [review]

**Comment by dgomezTT** [comment]
failed tests are the same than in main
